src/Adapter/MSTest.TestAdapter/Execution/TestExecutionManager.cs

```diff
@@ -314,37 +315,72 @@ private void ExecuteTestsInSource(IEnumerable<TestCase> tests, IRunContext? runC
                 }
 
                 var tasks = new List<Task>();
+                bool isSTA = false;
 
                 for (int i = 0; i < parallelWorkers; i++)
                 {
-                    tasks.Add(_taskFactory(() =>
+                    if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows)
+                        && MSTestSettings.RunConfigurationSettings.ExecutionApartmentState == ApartmentState.STA)
                     {
-                        while (!queue!.IsEmpty)
+                        isSTA = true;
+                        Thread entryPointThread = new(new ThreadStart(RunTest));
+
+                        entryPointThread.SetApartmentState(ApartmentState.STA);
+                        entryPointThread.Start();
+
+                        try
                         {
-                            if (_cancellationToken is { Canceled: true })
-                            {
-                                // if a cancellation has been requested, do not queue any more test runs.
-                                break;
-                            }
-
-                            if (queue.TryDequeue(out IEnumerable<TestCase>? testSet))
-                            {
-                                ExecuteTestsWithTestRunner(testSet, frameworkHandle, source, sourceLevelParameters, testRunner);
-                            }
+                            // should add cancellation toke??
+                            var threadTask = Task.Run(entryPointThread.Join);
+                            threadTask.Wait();
                         }
-                    }));
+                        catch (Exception ex)
+                        {
+                            RunCatch(ex);
+                            throw;
+                        }
+                    }
+                    else
+                    {
+                        tasks.Add(_taskFactory(() => RunTest()));
+                    }
                 }
 
-                try
+                if (!isSTA)
```

Comment:
I might be missing something but you are not waiting the end of execution of the threads (only the tasks). Given that you are wrapping threads into tasks, you should be able to always await the tasks.

Reply:
isn't that line a wait the thread to finish https://github.com/microsoft/testfx/pull/3213/files#diff-35b596abe791e1066c059e21cec09535297304887e34768fd2d7f3dc963eb65dR335

Reply:
It does yes, not sure why I missed that line! But the alg is then "broken" as your tests are not running in parallel, you will wait the end of execution of the thread before moving to the next iteration of the loop.